### PR TITLE
Add secure SSL support

### LINKED ISSUE #5923: Add support for SSL in Redshift connector
**Description:**
Implement SSL support in the Ballerina Redshift Connector to enable secure communication with Amazon Redshift databases.

To enable SSL in connections you have to set the `ssl` property to `true`. Redshift supports following `ssl modes`, which will be exposed in our ballerina clients. 
```
disable - SSL is disabled and the connection is not encrypted.

allow - SSL is used if the server requires it.

prefer - SSL is used if the server supports it. Amazon Redshift supports SSL, so SSL is used when you set sslmode to prefer.

require - SSL is required.

verify-ca - SSL must be used and the server certificate must be verified.

verify-full - SSL must be used. The server certificate must be verified and the server hostname must match the hostname attribute on the certificate.
```
**API Design**

```ballerina

# Possible values for the SSL mode.
public enum SSLMode {
    PREFER = "prefer",
    REQUIRE = "require",
    DISABLE = "disable",
    ALLOW = "allow",
    VERIFY_CA = "verify-ca",
    VERIFY_FULL = "verify-full"
}
```

Sample connection:

```ballerina
redshift:Options connectionOptions = {
       sslmode: redshift:VERIFY_FULL,
};

// Initialize the Redshift client
redshift:Client dbClient = check new (jdbcUrl, user, password, options = connectionOptions);
```

**Task Breakdown**
- [x] Design and implement the SSL support in Redshift Connector
- [x] Test positive use cases
- [ ] Test negative use cases
- [ ] Testing each ssl mode

**Referencecs:**
1. How to enable SSL in the server: https://www.intelligentdiscovery.io/controls/redshift/aws-redshift-enforce-ssl
2. Connecting with SSL: https://docs.aws.amazon.com/redshift/latest/mgmt/connecting-ssl-support.html

### PR DESCRIPTION
**Pull request by aashikam**
## Purpose

**Tested locally.**
**Ref:** https://docs.aws.amazon.com/redshift/latest/mgmt/connecting-ssl-support.html
fixes: https://github.com/ballerina-platform/ballerina-library/issues/5923

## Checklist
- [x] Linked to an issue
- [ ] Updated the specification
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Checked native-image compatibility

### CODE REVIEW

File: native/src/main/java/io/ballerina/lib/aws/redshift/Constants.java [hunk 1]
```diff
@@ -39,8 +39,18 @@ public static final class ClientConfiguration {
...
         public static final BString PROPERTIES = StringUtils.fromString("properties");
+        public static final BString SECURE_SOCKET = StringUtils.fromString("ssl");
     }
 
+    public static final class SSL {
+        public static final BString SSL_MODE = StringUtils.fromString("ssl");
+        public static final BString ROOT_CERT = StringUtils.fromString("rootcert");
+        public static final BString SSL_FACTORY_ARG = StringUtils.fromString("sslfactory");
+        public static final BString SSL_FACTORY_VALUE =
+                StringUtils.fromString("com.amazon.redshift.ssl.NonValidatingFactory");
```
Comment (Bhashinee) [inline native/src/main/java/io/ballerina/lib/aws/redshift/Constants.java]: Isn't this `NonValidatingFactory` class related to creating SSL connections without validating the server's SSL certificate?

File: ballerina/client.bal [hunk 2]
```diff
@@ -115,11 +115,20 @@ public isolated client class Client {
 #
 # + datasourceName - The driver class name to be used to get the connection
 # + properties - The database properties, which should be applied when getting the connection
+# + mode - Value for the SSL mode
 public type Options record {|
     string? datasourceName = ();
     map<anydata>? properties = ();
+    SSLMode sslMode = DISABLE;
```
Comment (ThisaruGuruge) [inline ballerina/client.bal]:
```suggestion
    SslMode sslMode = DISABLE;
```

File: ballerina/client.bal [hunk 3]
```diff
@@ -115,11 +115,20 @@ public isolated client class Client {
...
 |};
 
+# Possible values for the SSL mode.
+public enum SSLMode {
+    DISABLE = "disable",
+    VERIFY_CA = "verify-ca",
+    VERIFY_FULL = "verify-full"
```
Comment (ThisaruGuruge) [inline ballerina/client.bal]: Do we need to assign values for the enums here?

File: native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java [hunk 4]
```diff
@@ -83,33 +81,26 @@ public static Object createClient(BObject client, BMap<BString, Object> clientCo
...
-    // Unable to perform a complete validation since URL differs based on the database.
     private static boolean isJdbcUrlValid(String jdbcUrl) {
         return !jdbcUrl.isEmpty() && jdbcUrl.trim().startsWith("jdbc:");
     }
 
     public static Object close(BObject client) {
         return io.ballerina.stdlib.sql.nativeimpl.ClientProcessor.close(client);
     }
+
+    private static void addSSLOptions(BMap config, BMap<BString, Object> options) {
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java]:
```suggestion
    private static void addSslOptions(BMap config, BMap<BString, Object> options) {
```

File: native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java [hunk 5]
```diff
@@ -47,29 +47,27 @@ public static Object createClient(BObject client, BMap<BString, Object> clientCo
...
-                        poolProperties = new Properties();
-                        poolProperties.setProperty(Constants.POOL_CONNECTION_TIMEOUT,
-                                properties.getStringValue((BString) propKey).getValue());
+            if (options.getMapValue(Constants.ClientConfiguration.PROPERTIES) != null) {
+                properties = options.getMapValue(Constants.ClientConfiguration.PROPERTIES);
+                BString dataSourceNamVal = options.getStringValue(Constants.ClientConfiguration.DATASOURCE_NAME);
+                datasourceName = dataSourceNamVal == null ? null : dataSourceNamVal.getValue();
+                if (properties != null) {
+                    for (Object propKey : properties.getKeys()) {
+                        if (propKey.toString().toLowerCase(Locale.ENGLISH).matches(Constants.CONNECT_TIMEOUT)) {
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java]: What's the purpose of this check? I mean what's the purpose of using `toLowerCase()` here?

Another thing, if the `properties` is not null, the `propKey` must be of type `BString` right?

File: ballerina/client.bal (hunk 3 above)
Comment (aashikam) [reply ballerina/client.bal]: VERIFY_CA and VERIFY_FULL because of the underscore, can handle this internally also

**Review by Bhashinee — DISMISSED** [review]

**Comment by Bhashinee** [comment]
SSL part LGTM.

File: ballerina/client.bal (hunk 3 above)
Comment (ThisaruGuruge) [reply ballerina/client.bal]: I think it's better to handle this internally as we don't need to expose this information to the user.

File: native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java [hunk 6]
```diff
@@ -83,33 +81,26 @@ public static Object createClient(BObject client, BMap<BString, Object> clientCo
...
         return !jdbcUrl.isEmpty() && jdbcUrl.trim().startsWith("jdbc:");
     }
 
     public static Object close(BObject client) {
         return io.ballerina.stdlib.sql.nativeimpl.ClientProcessor.close(client);
     }
+
+    private static void addSSLOptions(BMap config, BMap<BString, Object> options) {
+        BString mode = config.getStringValue(Constants.SSL.SSL_MODE);
+        if (mode.getValue().equals(Constants.SSL.SSL_MODE_DISABLED.getValue())) {
```
Comment (Bhashinee) [inline native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java]: Can't we do the BString comparison without getting the String value of these two?

File: ballerina/client.bal [hunk 7]
```diff
@@ -115,18 +115,18 @@ public isolated client class Client {
...
 
 # Possible values for the SSL mode.
-public enum SSLMode {
-    DISABLE = "disable",
-    VERIFY_CA = "verify-ca",
-    VERIFY_FULL = "verify-full"
+public enum SslMode {
+    DISABLE,
+    VERIFY_CA,
+    VERIFY_FULL
```
Comment (Bhashinee) [inline ballerina/client.bal]: Shall we add doc comments to these three values as these appear in our API docs?

File: native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java [hunk 8]
```diff
@@ -83,33 +79,30 @@ public static Object createClient(BObject client, BMap<BString, Object> clientCo
...
         return io.ballerina.stdlib.sql.nativeimpl.ClientProcessor.close(client);
     }
+
+    private static void addSslOptions(BMap config, BMap<BString, Object> options) {
+        BString mode = config.getStringValue(Constants.SSL.SSL_MODE);
+        if (mode.equals(Constants.SSL.SSL_MODE_DISABLED)) {
+            options.put(Constants.SSL.SSL, false);
+        } else {
+            options.put(Constants.SSL.SSL, true);
+            if (mode.getValue().equals(Constants.SSL.SSL_MODE_VERIFY_CA.getValue())) {
```
Comment (Bhashinee) [inline native/src/main/java/io/ballerina/lib/aws/redshift/nativeimpl/ClientProcessor.java]:
```suggestion
            if (mode.equals(Constants.SSL.SSL_MODE_VERIFY_CA)) {
```

File: native/src/main/java/io/ballerina/lib/aws/redshift/Constants.java [hunk 9]
```diff
@@ -43,7 +43,10 @@ public static final class ClientConfiguration {
     public static final class SSL {
         public static final BString SSL = StringUtils.fromString("ssl");
         public static final BString SSL_MODE = StringUtils.fromString("sslMode");
-        public static final BString SSL_MODE_DISABLED = StringUtils.fromString("disable");
+        public static final BString SSL_MODE_DISABLED = StringUtils.fromString("DISABLE");
+        public static final BString SSL_MODE_VERIFY_CA = StringUtils.fromString("VERIFY_CA");
+        public static final BString SSL_MODE_VERIFY_CA_ARG = StringUtils.fromString("verify-ca");
+        public static final BString SSL_MODE_VERIFY_FULL_ARG = StringUtils.fromString("verify-full");
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/lib/aws/redshift/Constants.java:49]: We can define these as Java strings, right? We are not using them on the Ballerina side.

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**